File: backend/src/schemas/plan.py
from pydantic import BaseModel, Field
from typing import Literal
import uuid
from datetime import datetime

PlanStatus = Literal["draft", "generating", "active", "paused", "completed", "failed"]


class PlanConstraints(BaseModel):
    deadline_days: int | None = None
    team_size: int | None = None
    budget_usd: float | None = None
    tech_stack: list[str] = []
    notes: str | None = None


class PlanCreate(BaseModel):
    title: str = Field(min_length=1, max_length=200)
    goal: str = Field(min_length=10, max_length=2000)
    constraints: PlanConstraints = PlanConstraints()


class PlanUpdate(BaseModel):
    title: str | None = None
    goal: str | None = None
    status: PlanStatus | None = None
    constraints: PlanConstraints | None = None


class PlanOut(BaseModel):
    id: uuid.UUID
    title: str
    goal: str
    constraints: dict
    status: str
    risk_score: float | None
    confidence: float | None
    current_version: int
    job_id: str | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class PlanVersionOut(BaseModel):
    id: uuid.UUID
    plan_id: uuid.UUID
    version: int
    trigger: str
    snapshot: dict
    created_at: datetime

    model_config = {"from_attributes": True}


class DagOut(BaseModel):
    nodes: list[dict]
    edges: list[dict]
    critical_path: list[str]

File: frontend/src/types/plan.ts
export interface PlanConstraints {
  deadline_days?: number
  team_size?: number
  budget_usd?: number
  tech_stack?: string[]
  notes?: string
}

export interface Plan {
  id: string
  title: string
  goal: string
  constraints: PlanConstraints
  status: 'draft' | 'generating' | 'active' | 'paused' | 'completed' | 'failed'
  risk_score?: number
  confidence?: number
  current_version: number
  job_id?: string
  created_at: string
  updated_at: string
}

export interface PlanVersion {
  id: string
  plan_id: string
  version: number
  trigger: string
  snapshot: Record<string, unknown>
  created_at: string
}

export interface DagData {
  nodes: DagNode[]
  edges: DagEdge[]
  critical_path: string[]
}

export interface DagNode {
  id: string
  type: string
  data: {
    label: string
    category?: string
    status: string
    estimated_hours?: number
    priority: number
    is_on_critical_path: boolean
  }
  position?: { x: number; y: number }
}

export interface DagEdge {
  id: string
  source: string
  target: string
}
